src/psyclone/dynamo0p3.py

```diff
@@ -974,16 +979,33 @@ def _validate(self, need_evaluator):
                 "Kernel '{0}' specifies a gh_shape ({1}) but does not "
                 "need an evaluator because no basis or differential basis "
                 "functions are required".format(self.name, self._eval_shape))
-
-        # Check that this kernel only updates a single argument if an
-        # evaluator is required
-        if self._eval_shape == "gh_evaluator" and write_count > 1:
-            raise ParseError(
-                "A Dynamo 0.3 kernel requiring quadrature/evaluator must "
-                "only write to one argument but kernel {0} requires {1} and "
-                "updates {2} arguments".format(self.name,
-                                               self._eval_shape, write_count))
-
+        # Check that gh_evaluator_targets is only present if required
+        if self._eval_targets:
+            if not need_evaluator:
+                raise ParseError(
+                    "Kernel '{0}' specifies gh_evaluator_targets ({1}) but "
+                    "does not need an evaluator because no basis or "
+                    "differential basis functions are required".
+                    format(self.name, self._eval_targets))
+            if self._eval_shape != "gh_evaluator":
+                raise ParseError(
+                    "Kernel '{0}' specifies gh_evaluator_targets ({1}) but "
+                    "does not need an evaluator because gh_shape={2}".
+                    format(self.name, self._eval_targets, self._eval_shape))
+            # Check that there is a kernel argument on each of the
+            # specified spaces...
+            # Create a list (set) of the function spaces associated with
+            # the kernel arguments
+            fs_list = set()
+            for arg in self._arg_descriptors:
+                fs_list.update(arg.function_spaces)
+            # Check each evaluator_target against this list
+            for eval_fs in self._eval_targets:
+                if eval_fs not in fs_list:
+                    raise ParseError(
+                        "Kernel '{0}' specifies that an evaluator is required "
+                        "on {1} but does not have an argument on this space."
+                        .format(self.name, eval_fs))
```

Comment:
This is coming up against my limited understanding of what it is we're implementing. I *think* the FS metadata for basis/diff-basis functions is specifying something different to the ``evaluator_targets``. e.g. the former will specify that basis functions are required for the W0 space while the latter will specify which space(s) those basis functions must be evaluated on. @BShipway, could you confirm that that's correct?

Reply:
> This is coming up against my limited understanding of what it is we're implementing. I _think_ the FS metadata for basis/diff-basis functions is specifying something different to the `evaluator_targets`. e.g. the former will specify that basis functions are required for the W0 space while the latter will specify which space(s) those basis functions must be evaluated on. @BShipway, could you confirm that that's correct?

Yes. I think this is correct.  The idea is that the evaluator_targets specify the target dof locations at which the basis functions are evaluated.  E.g. we may have a kernel that writes to a W0 field and reads from a W3 field, but we might want to evaluate the W0 basis on the W3 points.  The evaluator_target for this would be [W3].